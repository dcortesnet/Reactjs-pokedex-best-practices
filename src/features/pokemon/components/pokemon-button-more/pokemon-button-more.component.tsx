import React, { useContext, useState } from "react";
import { PokemonContext } from "../../../../shared/context/pokemon.context";
import * as pokemonService from "../../../../shared/services/pokemon.service";
import { Pokemon } from "../../interfaces/pokemon.interface";
import { PokemonButtonMoreLayout } from "./pokemon-button-more.layout";

export const PokemonButtonMoreComponent = () => {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const { pokemon, setPokemon, nextPage, setNextPage } = useContext(
    PokemonContext
  );

  const getMorePokemon = async () => {
    try {
      setLoading(true);
      const page = await pokemonService.loadMorePokemon(nextPage);
      setNextPage(page.next);
      const pokemonByPage: Pokemon[] = await pokemonService.getAllPokemonByPage(
        page
      );
      setPokemon(pokemon.concat(pokemonByPage));
      setError(false);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  };

  return (
    <>
      {pokemon.length > 1 ? (
        <PokemonButtonMoreLayout
          loading={loading}
          onClick={getMorePokemon}
          error={error}
        ></PokemonButtonMoreLayout>
      ) : (
        ""
      )}
    </>
  );
};
