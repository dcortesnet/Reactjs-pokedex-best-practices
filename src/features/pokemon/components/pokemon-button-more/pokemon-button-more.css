.load-more {
  display: block;
  justify-content: center;
  text-align: center;
  width: 100%;
  padding-block-start: 50px;
  padding-block-end: 50px;
}

.button {
  margin-top: 15px;
  font-size: 1.5rem;
  border-radius: 5px;
  background-color: #dc3545;
  border-color: #dc3545;
  border: 1px solid transparent;
  color: #fff;
  padding: 20px;
}

.button:hover {
  border-color: brown;
  background: brown;
  cursor: pointer;
}

.button:focus {
  border-color: tomato;
  background: tomato;
}
