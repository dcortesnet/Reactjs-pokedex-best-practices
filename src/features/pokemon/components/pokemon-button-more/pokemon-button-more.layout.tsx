import React from "react";
import { ErrorMessageLayout } from "../../../../shared/components/error-message/error-message.layout";
import { LoadingSpinnerLayout } from "../../../../shared/components/loading-spinner/loading-spinner.layout";
import "./pokemon-button-more.css";

interface Props {
  loading: boolean;
  error: boolean;
  onClick: Function;
}

export const PokemonButtonMoreLayout: React.FC<Props> = ({
  loading,
  error,
  onClick,
}) => {
  return (
    <div>
      <div className="load-more">
        {loading ? (
          <LoadingSpinnerLayout />
        ) : error ? (
          <ErrorMessageLayout />
        ) : (
          <button className="button" onClick={() => onClick()}>
            Cargar m√°s pokemon
          </button>
        )}
      </div>
    </div>
  );
};
