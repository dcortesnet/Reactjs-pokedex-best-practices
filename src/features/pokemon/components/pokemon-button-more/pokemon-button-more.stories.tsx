import { Story, Meta } from "@storybook/react";
import { PokemonButtonMoreLayout } from "./pokemon-button-more.layout";

export default {
  title: "Pokemon/components/PokemonButtonMore",
  component: PokemonButtonMoreLayout,
} as Meta;

export const Template: Story<any> = (args) => <PokemonButtonMoreLayout {...args} />;
