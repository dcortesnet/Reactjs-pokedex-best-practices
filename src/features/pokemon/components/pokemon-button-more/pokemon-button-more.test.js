import { PokemonButtonMoreLayout } from "./pokemon-button-more.layout";
import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() });
describe("PokemonButtonMoreLayout", () => {
  it("Renders correctly", () => {
    const r = shallow(<PokemonButtonMoreLayout></PokemonButtonMoreLayout>);

    expect(toJson(r)).toMatchSnapshot();
  })

  it("Renders <button> and has 'Cargar m치s pokemon' text", () => {
    const text = "Cargar m치s pokemon";
    const r = shallow(<PokemonButtonMoreLayout></PokemonButtonMoreLayout>);

    expect(r.contains(text)).toEqual(true);

    expect(r.find("button").exists()).toBeTruthy();
  });

  it("Renders <LoadingSpinnerLayout>", () => {
    const r = shallow(
      <PokemonButtonMoreLayout loading></PokemonButtonMoreLayout>
    );

    expect(r.find("LoadingSpinnerLayout").exists()).toBeTruthy();
  });

  it("Renders <ErrorMessageLayout", () => {
    const r = shallow(
      <PokemonButtonMoreLayout error></PokemonButtonMoreLayout>
    );

    expect(r.find("ErrorMessageLayout").exists()).toBeTruthy();
  });

  it("Clicks button and has 'Cargar m치s pokemon' text", () => {
    const mockCallBack = jest.fn();
    const text = "Cargar m치s pokemon";
    const r = shallow(
      <PokemonButtonMoreLayout onClick={mockCallBack}></PokemonButtonMoreLayout>
    );

    expect(r.contains(text)).toEqual(true);

    r.find("button").simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
