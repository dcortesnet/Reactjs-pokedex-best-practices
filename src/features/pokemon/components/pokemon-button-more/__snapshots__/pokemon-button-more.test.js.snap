// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonButtonMoreLayout Renders correctly 1`] = `
<div>
  <div
    className="load-more"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Cargar m√°s pokemon
    </button>
  </div>
</div>
`;
